mozilla-detekt-rules/src/main/java/org/mozilla/fenix/detektrules/perf/MozillaNavigateCheck.kt

```diff
@@ -0,0 +1,51 @@
+package org.mozilla.fenix.detektrules.perf
+
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.resolve.calls.callUtil.getCall
+
+private const val VIOLATION_MSG = "Please use `org.mozilla.fenix.ext.NavController.navigateBlockingForAsyncNavGraph` " +
+        "instead because using `navigate` directly in the code can lead to the NavGraph not being loaded" +
+        "since it relies on a blocking call done in `navigateBlockingForAsyncNavGraph`."
+
+/**
+ * A check to prevent the use of `navController.navigate`. Since the NavGraph is loaded dynamically
+ * and asynchronously, there is a check in place to make sure the graph is loaded by wrapping the
+ * `navigate` function. However, using it directly might lead to code that needs the NavGraph being
+ * called before it being inflated.
+ */
+class MozillaNavigateCheck(config: Config = Config.empty) : Rule(config) {
+
+    override val issue = Issue(
+        "MozillaNavigateCheck",
+        Severity.Performance,
+        "Prevents us from calling `navController.navigate` instead of the functions that" +
+                "wrap it",
+        Debt.TEN_MINS
+    )
+
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        super.visitCallExpression(expression)
+
+        val calledMethod = expression.calleeExpression?.firstChild?.node?.chars
+
+        //We check for the navigate method and we have to ignore our extension function file, since
+        //we call navigate there
+        if (calledMethod == "navigate" && expression.containingFile.name != "NavController.kt") {
```

Comment:
We should suppress this with `@Suppress` in the file instead because it's more robust: it's tightly coupled to the file so we won't accidentally suppress in a file that happened to be named the same thing and it'll survive file renames.